Guard on inner tensor names for traceable wrapper subclasses

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129618

Fixes #129601

Background: it's possible that a traceable wrapper subclass will have an optional inner tensor constituent (e.g. NJT's cached min / max sequence lengths). To specify this, the subclass's `__tensor_flatten__()` impl should leave out any unspecified optional inner tensors in the returned list of `attrs`.

This PR guards on the list of inner tensor `attrs` returned in `subclass.__tensor_flatten__()[0]`.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

jbschlosser committed:
Guard on inner tensor names for traceable wrapper subclasses

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge -i

jbschlosser commented:
@pytorchbot merge -i

jbschlosser commented:
@pytorchbot merge -i

bdhirsh commented on torch/_dynamo/variables/builder.py:
```diff
@@ -1396,6 +1397,10 @@ def wrap_tensor(self, value: torch.Tensor):
         if is_traceable_wrapper_subclass(value):
             self.install_guards(GuardBuilder.TYPE_MATCH)
+            install_guard(
+                SubclassAttrListSource(source).make_guard(GuardBuilder.EQUALS_MATCH)
```
cc @tugsbayasgalan - we'll need to fix this for nested subclasses too. Dynamo generate a type match (and also am match on tensor_flatten attributes), and we should generate these guards recursively for inner tensors. (We can fix this independently of the metadata guard [issue](https://github.com/pytorch/pytorch/issues/114405))

jbschlosser committed:
Update on "Guard on inner tensor names for traceable wrapper subclasses"


Fixes #129601

Background: it's possible that a traceable wrapper subclass will have an optional inner tensor constituent (e.g. NJT's cached min / max sequence lengths). To specify this, the subclass's `__tensor_flatten__()` impl should leave out any unspecified optional inner tensors in the returned list of `attrs`.

This PR guards on the list of inner tensor `attrs` returned in `subclass.__tensor_flatten__()[0]`.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge -i
